core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -199,82 +173,75 @@ private synchronized void putClearModification() {
+      }
 
-               if (!ourModificationsToReplicate.isEmpty()) {
-                  asyncBatchStage = asyncBatchStage.thenCompose(ignore -> {
-                     if (trace) {
-                        log.tracef("Sending batch write/remove operations %s to underlying store with id %s", ourModificationsToReplicate.values(),
-                              System.identityHashCode(ourModificationsToReplicate));
-                     }
-                     return retry(() -> replicateModifications(ourModificationsToReplicate), persistenceConfiguration.connectionAttempts()).whenComplete((ignore2, t) -> {
-                        synchronized (this) {
-                           replicatingModifications = null;
-                        }
-                     });
-                  });
+      if (!ourModificationsToReplicate.isEmpty()) {
+         asyncBatchStage = asyncBatchStage.thenCompose(ignore -> {
+            if (trace) {
+               log.tracef("Sending batch write/remove operations %s to underlying store with id %s", ourModificationsToReplicate.size(),
+                     System.identityHashCode(ourModificationsToReplicate));
+            }
+            return retry(() -> replicateModifications(ourModificationsToReplicate), persistenceConfiguration.connectionAttempts()).whenComplete((ignore2, t) -> {
+               synchronized (this) {
+                  replicatingModifications = null;
                }
-
-               asyncBatchStage.whenComplete((ignore, t) -> {
-                  if (trace) {
-                     log.tracef("Async operations completed for id %s", System.identityHashCode(ourModificationsToReplicate));
-                  }
-                  boolean submitNewBatch;
-                  CompletableFuture<Void> future;
-                  synchronized (this) {
-                     submitNewBatch = !pendingModifications.isEmpty() || hasPendingClear;
-                     future = batchFuture;
-                     batchFuture = submitNewBatch ? new CompletableFuture<>() : null;
-                  }
-                  if (t != null) {
-                     future.completeExceptionally(t);
-                  } else {
-                     future.complete(null);
-                  }
-                  if (submitNewBatch) {
-                     if (trace) {
-                        log.trace("Submitting new batch after completion of prior");
-                     }
-                     requestFlowable.onNext(requestFlowable);
-                  }
-               });
             });
+         });
+      }
+
+      asyncBatchStage.whenComplete((ignore, t) -> {
+         if (trace) {
+            log.tracef("Async operations completed for id %s", System.identityHashCode(ourModificationsToReplicate));
+         }
+         boolean submitNewBatch;
+         CompletableFuture<Void> future;
+         synchronized (this) {
+            submitNewBatch = !pendingModifications.isEmpty() || hasPendingClear;
```

Comment:
The assignment of `batchFuture` protects that. You can only submit a batch if you assigned it to a new value.

Reply:
I'm pretty sure the assignment of `batchFuture` protects agains two threads both seeing a non-empty `pendingModifications` and each assigning its own `CompletableFuture` to `batchFuture`. 

OTOH I'm sure there's something else preventing 2 threads from both submitting a batch, it's just not clear from the code here.

Reply:
You can only submit a batch if you wrote to `batchFuture` though.

Reply:
writing to `batchFuture` and submitting the batch are tied together and should only change together.

Reply:
Okay, I understand your concern. The part you are missing is that it isn't possible for two threads to invoke this to begin with as only a single thread can invoke `submitTask` at a time at https://github.com/infinispan/infinispan/blob/6729d3bca7492b317e5ce0d8a6e1c2b8fec93641/core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java#L476.

So with this only one thread in `submitModitification` can assign the `batchFuture` and only if it is null. This means that `submitTask` is invoked in a thread at a time and `batchFuture` will be non null at the beginning, so its assignment you linked here will either put it to null or a new non null value. In the former case the loop doesn't continue, otherwise it resubmits the task again.

I can add a more explicit comment to `submitTask` detailing this. It had a very brief comment, but wasn't really detailing this.

Reply:
It's still not very clear to me, I'll definitely need a more explicit comment :)

So you're saying that when this code executes, `batchFuture` must be non-null, and any `submitTask()` running on another thread cannot submit a new batch?

Reply:
> It's still not very clear to me, I'll definitely need a more explicit comment :)
> 
> So you're saying that when this code executes, `batchFuture` must be non-null, and any `submitTask()` running on another thread cannot submit a new batch?

No, as mentioned on the `submitTask` method it states that `submitModification` checks the variable and ensures that only single modification can start the call. So it isn't possible for `submitTask` to be invoked from multiple threads concurrently. I have added the latter part to the method description as well.

Reply:
It would be simpler if `submitModification()` was the only caller of `submitTask()`.

But in `submitTask()` itself you have

```
asyncBatchStage.whenComplete((ignore, t) -> {
...
            submitTask();
         }
```

Reply:
Which is why I mentioned `Invoking this method outside of this method...` in the `submitTask` javadoc.

But the `submitTask` method is the only method that ever assigns `batchFuture` to null, which prevents concurrent modification as well. I have added another paragraph to the Javadoc.

Reply:
> Which is why I mentioned `Invoking this method outside of this method...` in the `submitTask` javadoc.

Sorry, that was too subtle for me... I guess my eyes just glazed over the second "this method".

> But the `submitTask` method is the only method that ever assigns `batchFuture` to null, which prevents concurrent modification as well. I have added another paragraph to the Javadoc.

Part of the reason I didn't read the javadoc correctly is that I didn't see the `whenComplete()` lambda as part of "this method". Lexically it's a part of `submitTask()`, but at the JVM level it's a separate method, and it runs separately, not during the `submitTask()` invocation.